src/api/common/components/ComponentService.ts

```diff
@@ -0,0 +1,228 @@
+import { type BaseComponentEntity } from '@common/components/BaseComponentEntity.js';
+import { AppDataSource } from '@config/connection.js';
+import { getComponentMetadata, type ComponentMetadata } from '@decorators/StrapiComponent.js';
+import { In } from 'typeorm';
+
+export class ComponentService {
+  static async transformComponents<T>(entity: T, entityClass: Function, requestedRelations?: string[]): Promise<T> {
+    const entityAny = entity as any;
+
+    if (!entityAny.components || entityAny.components.length === 0) {
+      delete entityAny.components;
+      return entity;
+    }
+
+    const metadata = getComponentMetadata(entityClass);
+    if (metadata.length === 0) {
+      delete entityAny.components;
+      return entity;
+    }
+
+    const filteredMetadata = requestedRelations
+      ? metadata.filter(
+          meta => requestedRelations.includes(meta.field) || requestedRelations.includes(meta.propertyKey),
+        )
+      : metadata;
+
+    await Promise.all(
+      filteredMetadata.map(async (meta: ComponentMetadata) => {
+        const componentRecord = (entityAny.components as BaseComponentEntity[]).find(
+          cmp => cmp.field === meta.field && cmp.componentType === meta.componentType,
+        );
+
+        if (componentRecord) {
+          const repo = AppDataSource.getRepository(meta.entity);
+          const data = await repo.findOne({ where: { id: componentRecord.cmpId } });
+          entityAny[meta.propertyKey] = data;
+        } else {
+          entityAny[meta.propertyKey] = null;
+        }
+      }),
+    );
+
+    delete entityAny.components;
+    return entity;
+  }
+
+  static async transformMultipleComponents<T>(
+    entities: T[],
+    entityClass: Function,
+    requestedRelations?: string[],
+  ): Promise<T[]> {
+    if (entities.length === 0) {
+      return entities;
+    }
+
+    const metadata = getComponentMetadata(entityClass);
+    if (metadata.length === 0) {
+      return entities;
+    }
+
+    const filteredMetadata = requestedRelations
+      ? metadata.filter(
+          meta => requestedRelations.includes(meta.field) || requestedRelations.includes(meta.propertyKey),
+        )
+      : metadata;
+
+    const componentCache = new Map<string, Map<number, any>>();
+
+    for (const meta of filteredMetadata) {
+      const componentIds: number[] = [];
+
+      for (const entity of entities) {
+        const entityAny = entity as any;
+        if (!entityAny.components || entityAny.components.length === 0) continue;
+
+        const componentRecord = (entityAny.components as BaseComponentEntity[]).find(
+          cmp => cmp.field === meta.field && cmp.componentType === meta.componentType,
```

Comment:
El objeto **`cmp`** asumo que significa **`component`**.

Pero creo que podría ser mas descriptiva su definición.

```suggestion
          component => component.field === meta.field && component.componentType === meta.componentType,
```